rfcs/0014-opendal.md

```diff
@@ -0,0 +1,196 @@
+
+1. **Caching**: We implemented `CachedObjectStore` as a wrapper around `object_store::ObjectStore` to provide transparent chunked caching. This required us to maintain our own caching logic and handle cache invalidation strategies.
+
+2. **Retrying**: We implemented `RetryingObjectStore` as another wrapper layer to automatically retry operations on transient failures. This added complexity in determining which errors are retryable and managing backoff strategies.
+
+These wrapper implementations have several drawbacks:
+
+- Each feature requires implementing the full `ObjectStore` trait, leading to significant boilerplate code
+- We're duplicating efforts that are already solved problems in the broader ecosystem
+- Maintaining and testing these wrappers adds ongoing engineering overhead
+- The composition of multiple wrapper layers can be error-prone and harder to reason about
+
+There were issues that discussed about introducing some composability for `object_store` crate:
+
+- https://github.com/apache/arrow-rs-object-store/issues/14
+- https://github.com/apache/arrow-rs-object-store/issues/274
+
+Unfortunately, these issues still have little progress.
+
+We hope to have a solution that reduces the complexity of maintaining our own wrapper layers while leveraging extension capabilities from an actively developed ecosystem.
+
+**OpenDAL** offers a promising solution: it's an Apache Incubator project that provides a **composable layer system** where features like caching, retrying, and observability are production-ready layers that can be combined without custom wrapper code. This addresses our exact pain points while being actively maintained by a nice community.
+
+<!-- TOC --><a name="goals-and-constraints"></a>
+## Goals and Constraints
+
+<!-- TOC --><a name="goals"></a>
+### Goals
+
+The primary goals of migrating to OpenDAL are:
+
+1. **Reduce maintenance burden**: Eliminate the need to maintain custom wrapper layers (`CachedObjectStore`, `RetryingObjectStore`) by leveraging battle-tested ecosystem solutions
+2. **Improve extensibility**: Gain access to OpenDAL's rich layer ecosystem for future capabilities (observability, rate limiting, etc.) without additional development effort
+3. **Future-proof storage integration**: Benefit from OpenDAL's active development and broader backend support as SlateDB evolves
+
+<!-- TOC --><a name="constraints"></a>
+### Constraints
+
+Given that `object_store` is a core dependency deeply integrated throughout SlateDB, we must ensure:
+
+1. **No breaking changes**: The migration must be transparent to SlateDB users. Since we currently expose the `object_store::ObjectStore` trait as our public I/O API, we need to maintain API compatibility—either by keeping the same trait interface with a compatibility layer that preserves existing behavior.
```

Comment:
Nit: "either by keeping the same"... there's no "or" here. What's the alternative?